wordpress-mobile/WordPress-Android
# Async media: Publish/Save while uploading media

Adds support for publishing/saving posts while they have media uploads in progress.

#### Caveats:
1. Doesn't handle cancelled images (this could only happen if the `MediaUploadService` were destroyed while uploading media) - will be handled in a future PR
2. There's no messaging of what's going on (other than the 'Uploading...' state of the post card in the list) until the post upload completes and we see a notification. The upload UI is getting an overhaul as part of the async experience (in future PRs). In the meantime, I'll open a separate PR adding plain notifications for async uploads for demonstration purposes.
3. This breaks reattachment, in the sense that you can't re-open a post once you've exited the editor, until it finishes uploading

#### To test:

A:
1. Create or edit a post, and add media
2. While media are uploading, press Publish (or Update)
3. The post should be in 'Uploading' state in the post list
4. Wait for the post to finish uploading (the state will update, and you'll get a 'Published' notification)
5. Open the post, and confirm that all media are uploaded and not local

B:
1. Create or edit a post, and add media
2. While media are uploading, **exit the Editor**
3. The post should be in 'Uploading' state in the post list
4. Wait for the post to finish uploading (the state will update to `Draft`, and you'll get a 'Published' notification)
5. Open the post, and confirm that all media are uploaded and not local

C:
1. Repeat A and B, but induce an error in the image upload
2. The post should be saved as a local draft
3. Opening the post should have the failed media in 'failed' status

Also check that uploading/closing the editor with already failed media behaves the same as before, and that non-media uploads work correctly.

cc @mzorz and @daniloercoli
